From 764053ba137355f67a62c2b6989aed68cbddf17a Mon Sep 17 00:00:00 2001
From: Chau Ly <chaul@amperecomputing.com>
Date: Mon, 5 May 2025 03:28:40 +0000
Subject: [PATCH] platform-mc: Parse File Descriptor PDR

This commit supports parsing File Descriptor PDRs for each terminus,
and build the file information map for that terminus. This map also
holds information about the organization of the files, represented
through the file path hierarchy. The map will later be used to
construct file objects on D-Bus.

Upstream-Status: Submitted [https://gerrit.openbmc.org/c/openbmc/pldm/+/82342]

Change-Id: I5d602ba732065a47874280d71f8c38860119b041
Signed-off-by: Chau Ly <chaul@amperecomputing.com>
(cherry picked from commit 14051a68055765bc06ead48dc7c13e92fe5da442)
---
 common/types.hpp         |  22 +++++
 platform-mc/terminus.cpp | 193 +++++++++++++++++++++++++++++++++++++++
 platform-mc/terminus.hpp |  38 ++++++++
 3 files changed, 253 insertions(+)

diff --git a/common/types.hpp b/common/types.hpp
index 298f426..70fe09d 100644
--- a/common/types.hpp
+++ b/common/types.hpp
@@ -285,4 +285,26 @@ using Callback = std::function<void()>;
 
 } // namespace bios
 
+namespace file_transfer
+{
+using FileID = uint16_t;
+
+enum class FileClassification : uint8_t
+{
+    OEM = 0x00,
+    BootLog,
+    SerialTxFIFO,
+    SerialRxFIFO,
+    DiagnosticLog,
+    CrashDumpFile,
+    SecurityLog,
+    FRUDataFile,
+    TelemetryDataFile,
+    TelemetryDataLog,
+    OtherLog = 0xFD,
+    OtherFile = 0xFE,
+    FileDirectory = 0xFF,
+};
+} // namespace file_transfer
+
 } // namespace pldm
diff --git a/platform-mc/terminus.cpp b/platform-mc/terminus.cpp
index 0ea4135..78d850e 100644
--- a/platform-mc/terminus.cpp
+++ b/platform-mc/terminus.cpp
@@ -10,12 +10,16 @@
 
 #include <memory>
 #include <ranges>
+#include <regex>
 
 namespace pldm
 {
 namespace platform_mc
 {
 
+static const std::regex objPathAntiPattern("[^a-zA-Z0-9_/]+");
+static constexpr FileID virtualTopMostFileId = 0xFFFF;
+
 Terminus::Terminus(pldm_tid_t tid, uint64_t supportedTypes,
                    sdeventplus::Event& event) :
     initialized(false), maxBufferSize(PLDM_PLATFORM_EVENT_MSG_MAX_BUFFER_SIZE),
@@ -221,6 +225,20 @@ void Terminus::parseTerminusPDRs()
                 entityAuxiliaryNamesTbl.emplace_back(std::move(entityNames));
                 break;
             }
+            case PLDM_FILE_DESCRIPTOR_PDR:
+            {
+                auto parsedPdr = parseFileDescriptorPDR(pdr);
+                if (!parsedPdr)
+                {
+                    lg2::error(
+                        "Failed to parse PDR with type {TYPE} handle {HANDLE}",
+                        "TYPE", pdrHdr->type, "HANDLE",
+                        static_cast<uint32_t>(pdrHdr->record_handle));
+                    continue;
+                }
+                fileDescriptorPdrs.emplace_back(std::move(parsedPdr));
+                break;
+            }
             default:
             {
                 lg2::error("Unsupported PDR with type {TYPE} handle {HANDLE}",
@@ -277,6 +295,10 @@ void Terminus::parseTerminusPDRs()
 #endif
     }
 
+    if (!fileDescriptorPdrs.empty())
+    {
+        buildFileHierarchy();
+    }
     addNextSensorFromPDRs();
 }
 
@@ -998,5 +1020,176 @@ int Terminus::fetchEntityManagerConfiguration()
     return 0;
 }
 
+std::shared_ptr<pldm_platform_file_descriptor_pdr>
+    Terminus::parseFileDescriptorPDR(const std::vector<uint8_t>& pdrData)
+{
+    if (pdrData.empty())
+    {
+        lg2::error("Failed to parse File Descriptor PDR data of empty size.");
+        return nullptr;
+    }
+    const uint8_t* ptr = pdrData.data();
+    auto parsedPdr = std::make_shared<pldm_platform_file_descriptor_pdr>();
+    auto rc = decode_pldm_platform_file_descriptor_pdr(ptr, pdrData.size(),
+                                                       parsedPdr.get());
+    if (rc)
+    {
+        lg2::error("Failed to decode File Descriptor PDR data, error {RC}.",
+                   "RC", rc);
+        return nullptr;
+    }
+    return parsedPdr;
+}
+
+void Terminus::constructFilePathRecursive(
+    std::shared_ptr<FileInfoStruct> file,
+    const sdbusplus::message::object_path& basePath)
+{
+    if (!file)
+    {
+        lg2::error("FileInfoStruct object pointer is invalid for "
+                   "construction of a file path under {BASEPATH}.",
+                   "BASEPATH", basePath);
+        return;
+    }
+
+    if ((file->name).empty())
+    {
+        lg2::error("FileInfoStruct name is empty for "
+                   "construction of a file path under {BASEPATH}.",
+                   "BASEPATH", basePath);
+        return;
+    }
+
+    file->objPath = basePath;
+
+    std::string cleanName =
+        std::regex_replace(file->name, objPathAntiPattern, "_");
+    file->objPath /= cleanName;
+
+    if ((file->children).empty())
+    {
+        return;
+    }
+
+    for (auto& child : file->children)
+    {
+        constructFilePathRecursive(child, file->objPath);
+    }
+}
+
+void Terminus::buildFileHierarchy()
+{
+    if (terminusName.empty())
+    {
+        lg2::error(
+            "Terminus ID {TID}: DOES NOT have name. Skip adding File Descriptor.",
+            "TID", tid);
+        return;
+    }
+
+    if (fileDescriptorPdrs.empty())
+    {
+        return;
+    }
+
+    /*
+     * DSP0248 v1.3.0 Table 108 – File Descriptor PDR specifies that the
+     * SuperiorDirectoryFileIdentifier field in the File Descriptor PDR is
+     * 0xFFFF when
+     * 1. If the EntityType is a Device File, then this PDR is not part of any
+     * hierarchy
+     * 2. If the EntityType is a Device File Directory, then this PDR is a top
+     * most element of the hierarchy.
+     *
+     * Therefore, 0xFFFF is chosen as the FileIdentifier of the top most
+     * element. But as FileIdentifier is unique only within a terminus, so the
+     * top most element should be the terminus itself. As a result, a virtual
+     * top most element aka the terminus will be constructed to the file info
+     * map, using 0xFFFF as its key and its parent. Any Device File Directory or
+     * Device File that has an 0xFFFF SuperiorDirectoryFileIdentifier will be
+     * placed right under this top most element in the hirerachy.
+     *
+     * The involvement of EAR in representing the hierarchy is not supported
+     * yet.
+     */
+    std::map<FileID, std::shared_ptr<FileInfoStruct>> fileInfoMap;
+    auto topMostInfo = std::make_shared<FileInfoStruct>(
+        FileInfoStruct{terminusName, virtualTopMostFileId, {}, nullptr, {}});
+    fileInfoMap.emplace(virtualTopMostFileId, std::move(topMostInfo));
+    // PASS 1: Construct initial file info
+    for (const auto& pdr : fileDescriptorPdrs)
+    {
+        if (pdr->file_name.length == 0)
+        {
+            lg2::error("Skipping addition of file identifier {ID} to file "
+                       "info map because of an 0 file name length.",
+                       "ID", pdr->file_identifier);
+            continue;
+        }
+        // Exclude the NULL terminator character at the end of ASCII string
+        std::string fileName((const char*)(pdr->file_name.ptr),
+                             pdr->file_name.length - 1);
+
+        if (fileName.empty())
+        {
+            lg2::error("Skipping addition of file identifier {ID} to file "
+                       "info map because of an empty file name.",
+                       "ID", pdr->file_identifier);
+            continue;
+        }
+        lg2::info("Adding file {FILE} to file info map.", "FILE", fileName);
+        auto fileInfo = std::make_shared<FileInfoStruct>(FileInfoStruct{
+            fileName, pdr->superior_directory_file_identifier, {}, pdr, {}});
+        fileInfoMap.emplace(pdr->file_identifier, std::move(fileInfo));
+    }
+
+    // PASS 2: Add children files to their parent info
+    for (const auto& file : fileInfoMap)
+    {
+        const auto& parentId = file.second->parentId;
+        // Avoid adding top most file as its child
+        if (fileInfoMap.contains(parentId) && (parentId != file.first))
+        {
+            fileInfoMap[parentId]->children.push_back(file.second);
+        }
+    }
+
+    // Recursively construct file hierarchy
+    const sdbusplus::message::object_path basePath(
+        "/xyz/openbmc_project/pldm/file/");
+    auto& topMostFile = fileInfoMap[virtualTopMostFileId];
+    constructFilePathRecursive(topMostFile, basePath);
+
+    for (const auto& file : fileInfoMap)
+    {
+        if (file.first == virtualTopMostFileId)
+        {
+            // Don't add the virtual top most file
+            continue;
+        }
+
+        const auto& fileInfo = file.second;
+
+        if (std::string(fileInfo->objPath).empty())
+        {
+            lg2::error("Skipping object creation of File Descriptor {NAME} "
+                       "because of an empty object path",
+                       "NAME", fileInfo->name);
+            continue;
+        }
+
+        if (static_cast<FileClassification>(
+                fileInfo->pdr->file_classification) ==
+            FileClassification::FileDirectory)
+        {
+            // Don't add directories
+            continue;
+        }
+
+        lg2::debug("Constructing file descriptor object {PATH}.", "PATH",
+                   file.second->objPath);
+    }
+}
 } // namespace platform_mc
 } // namespace pldm
diff --git a/platform-mc/terminus.hpp b/platform-mc/terminus.hpp
index 00ce5f2..2b8ef56 100644
--- a/platform-mc/terminus.hpp
+++ b/platform-mc/terminus.hpp
@@ -22,6 +22,18 @@ namespace platform_mc
 {
 
 using namespace pldm::pdr;
+using namespace pldm::file_transfer;
+
+struct FileInfoStruct
+{
+    std::string name; //!< File name
+    FileID parentId;  //!< FileIdentifier of the parent file
+    sdbusplus::message::object_path objPath; //!< File object path
+    std::shared_ptr<pldm_platform_file_descriptor_pdr>
+        pdr;                                 //!< Pointer to file descriptor pdr
+    std::vector<std::shared_ptr<FileInfoStruct>>
+        children;                            //!< Vector of child files
+};
 
 /**
  * @brief Terminus
@@ -319,6 +331,28 @@ class Terminus
      */
     void addNextSensorFromPDRs();
 
+    /** @brief Parse the File Descriptor PDRs
+     *
+     *  @param[in] pdrData - the response PDRs from GetPDR command
+     *
+     *  @return pointer to File Descriptor info struct
+     */
+    std::shared_ptr<pldm_platform_file_descriptor_pdr> parseFileDescriptorPDR(
+        const std::vector<uint8_t>& pdrData);
+
+    /** @brief Recursively construct file object path for the file info map
+     *
+     *  @param[in] file - pointer to the current file in the file tree
+     *  @param[in] basePath - the parent object path of the current file
+     */
+    void constructFilePathRecursive(
+        std::shared_ptr<FileInfoStruct> file,
+        const sdbusplus::message::object_path& basePath);
+
+    /** @brief Build file info map for the terminus
+     */
+    void buildFileHierarchy();
+
     /* @brief The terminus's TID */
     pldm_tid_t tid;
 
@@ -377,6 +411,10 @@ class Terminus
     std::vector<std::shared_ptr<pldm_compact_numeric_sensor_pdr>>
         compactNumericSensorPdrs{};
 
+    /** @brief File Descriptor PDR list */
+    std::vector<std::shared_ptr<pldm_platform_file_descriptor_pdr>>
+        fileDescriptorPdrs{};
+
     /** @brief Iteration to loop through sensor PDRs when adding sensors */
     SensorID sensorPdrIt = 0;
 };
